pkg/autoupdate/tuf/library_lookup.go

```diff
@@ -0,0 +1,75 @@
+package tuf
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/kolide/launcher/pkg/autoupdate"
+)
+
+// CheckOutLatest returns the path to the latest downloaded executable for our binary, as well
+// as its version.
+func CheckOutLatest(binary autoupdatableBinary, rootDirectory string, updateDirectory string, channel string) (string, string, error) {
+	if updateDirectory == "" {
+		updateDirectory = defaultLibraryDirectory(rootDirectory)
+	}
+
+	releasePath, releaseVersion, err := findExecutableFromRelease(binary, LocalTufDirectory(rootDirectory), channel, updateDirectory)
+	if err == nil {
+		return releasePath, releaseVersion, nil
+	}
+
+	// If we can't find the specific release version that we should be on, then just return the executable
+	// with the most recent version in the library
+	return mostRecentVersion(binary, updateDirectory)
+}
+
+// findExecutableFromRelease looks at our local TUF repository to find the release for our
+// given channel. If it's already downloaded, then we return its path and version.
+func findExecutableFromRelease(binary autoupdatableBinary, tufRepositoryLocation string, channel string, baseUpdateDirectory string) (string, string, error) {
+	// Initialize a read-only TUF metadata client to parse the data we already have downloaded about releases.
+	metadataClient, err := readOnlyTufMetadataClient(tufRepositoryLocation)
+	if err != nil {
+		return "", "", errors.New("could not initialize TUF client, cannot find release")
+	}
+
+	// From already-downloaded metadata, look for the release version
+	targets, err := metadataClient.Targets()
+	if err != nil {
+		return "", "", fmt.Errorf("could not get target: %w", err)
+	}
+
+	targetName, _, err := findRelease(binary, targets, channel)
+	if err != nil {
+		return "", "", fmt.Errorf("could not find release: %w", err)
+	}
+
+	targetPath, targetVersion := pathToTargetVersionExecutable(binary, targetName, baseUpdateDirectory)
+	if autoupdate.CheckExecutable(context.TODO(), targetPath, "--version") != nil {
+		return "", "", fmt.Errorf("version %s from target %s either not yet downloaded or corrupted: %w", targetVersion, targetName, err)
+	}
+
+	return targetPath, targetVersion, nil
+}
+
+// mostRecentVersion returns the path to the most recent, valid version available in the library for the
+// given binary, along with its version.
+func mostRecentVersion(binary autoupdatableBinary, baseUpdateDirectory string) (string, string, error) {
```

Comment:
consider naming return values

Reply:
I agree that more than 2 return values is confusing, but I think naming them is an antipattern. (Though there's disagreement in go)

It looks pretty, but it creates them as implicit vars, which always feels like it breaks scoping to me

Reply:
eg: I just accept it as imperfect. Since it feels least bad. We could talk about changing that?

Reply:
Discussed offline, updated to return a struct